cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -0,0 +1,145 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <chrono>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/status.h"
+
+namespace arrow {
+
+namespace io {
+
+class InputStream;
+class OutputStream;
+
+}  // namespace io
+
+namespace fs {
+
+// A system clock time point expressed as a 64-bit (or more) number of
+// secondes since the epoch.
+using TimePoint = std::chrono::sys_time<std::chrono::nanoseconds>;
+
+enum class FileType {
+  // Target does not exist
+  NonExistent,
+  // Target exists but its type is unknown (could be a special file such
+  // as a Unix socket or character device, or Windows NUL / CON / ...)
+  Unknown,
+  // Target is a regular file
+  File,
+  // Target is a directory
+  Directory
+};
+
+std::string ToString(FileType);
+
+struct FileStats {
+  FileType type = FileType::Unknown;
+  // The file path in the filesystem.
+  std::string path;
+  // The size in bytes, if available.  Only regular files are guaranteed
+  // to have a size.
+  int64_t size = -1;
+  // The time of last modification, if available.
+  TimePoint mtime = TimePoint(-1);
+
+  FileStats() = default;
+
+  // The file base name (component after the last '/').
+  std::string base_name() const;
+};
+
+struct OpenFlags {
+  // Whether to transparently compress or decompress, based on file extension
+  // or other heuristics.
+  bool transparent_compression = true;
+
+  OpenFlags() = default;
+};
+
+struct Selector {
+  // The folder in which to select files.
+  std::string base_folder;
+  // Whether to recurse into subfolders.
+  bool recursive = false;
+
+  MatchOptions() = default;
+};
+
+
+class FileSystem {
+ public:
+  virtual ~FileSystem();
+
+  /// Get statistics for the given target.
+  ///
+  /// Any symlink is automatically dereferenced, recursively
+  virtual Status GetTargetStats(const std::string& path, FileStats* out);
+  /// Same, for many targets at once.
+  virtual Status GetTargetStats(const std::vector<std::string>& paths,
+                                std::vector<FileStats>* out) = 0;
+  virtual Status GetTargetStats(const Selector& select,
+                                std::vector<FileStats>* out) = 0;
+
+  /// Create a folder and subfolders.
+  virtual Status CreateFolder(const std::string& path, bool recursive = true) = 0;
+
+  /// Delete a folder and contents.
+  virtual Status DeleteFolder(const std::string& path) = 0;
+
+  /// Delete a file.
+  virtual Status DeleteFile(const std::string& path) = 0;
+  /// Delete many files.
+  virtual Status DeleteFiles(const std::vector<string>& path) = 0;
+
+  /// Move / rename a file or directory.
+  ///
+  /// The destination will be replaced if it exists.
+  virtual Status Move(const std::string& src, const std::string& dest) = 0;
```

Comment:
Could call this `RenameFile` (as TF does) instead of `Move`. I note that `Move` operations on directories may not be natively supported and may have to be emulated for e.g. S3

Reply:
I don't know. `Move` implies that the destination could be clobbered. `Rename` also implies you're staying in the same directory.